crates/noirc_evaluator/src/ssa/ir/function_inserter.rs

```diff
@@ -47,12 +47,18 @@ impl<'f> FunctionInserter<'f> {
 
     /// Insert a key, value pair if the key isn't already present in the map
     pub(crate) fn try_map_value(&mut self, key: ValueId, value: ValueId) {
-        self.values.entry(key).or_insert(value);
+        if key != value {
+            self.values.entry(key).or_insert(value);
+        }
     }
 
     /// Insert a key, value pair in the map
     pub(crate) fn map_value(&mut self, key: ValueId, value: ValueId) {
-        self.values.insert(key, value);
+        if key == value {
+            self.values.remove(&key);
```

Comment:
So for `map_value` we remove the key if `key == value`, but in try_map_value we don't attempt to remove if `key == value`. Could you elaborate a bit more on this reasoning?

Reply:
We remove the key if `key == value` since if `key` isn't in the map in `resolve`, we default to the key anyway. Previously mapping key to itself was valid and did the same thing explicitly, but now that we resolve recursively it'd lead to an infinite loop.

`try_map_value` only does something if the key isn't already in the map, so if the key isn't already in the map and `key == value`, we want to remove it normally but we already know it isn't in the map

Reply:
Thanks that makes sense

Reply:
I'll add a comment explaining this

Reply:
I ended up just adding the case with a small comment. It is technically unneeded but we shouldn't hit that case anyway and it makes the code easier to understand.